### PR TITLE
10bit ii comp atb

### PR DESCRIPTION
**Pull request by okhlif**
### Description

- Address and enable Compound Inter-Intra/Compound Inter-Inter and ATB for 10bit MD
- Fix ATB for 8bit and 10bit MD
Associated to issue #622 

###  Author
@okhlif 

### Type of change
Added feature
Bug fix

### Tests and performance

**Speed on cloud Instance [720p 10bit]** 

Encoder |  dev%
-- | --
hbd 0 (ii/Comp/ATB) ON vs OFF   (reference) | 58%
hbd 1 (ii/Comp/ATB)  ON vs OFF | 54%

**BDR 10bit**

360p   10bit | Y PSNR | U PSNR | V PSNR | Y SSIM | U SSIM | V SSIM
-- | -- | -- | -- | -- | -- | --
hbd 0   (ii/Comp/ATB) ON vs OFF (reference) | -1.3% | -1.8% | -1.1% | -1.1% | -1.9% | -1.7%
hbd 1   (ii/Comp/ATB)  ON vs OFF | -2.5% | -2.5% | -2.1% | -2.6% | -2.9% | -3.6%

### CODE REVIEW

File: Source/App/EncApp/EbAppConfig.c [hunk 1]
```diff
@@ -546,7 +546,7 @@ void eb_config_ctor(EbConfig *config_ptr)
     config_ptr->hme_level2_search_area_in_height_array[0]  = 1;
     config_ptr->hme_level2_search_area_in_height_array[1]  = 1;
     config_ptr->screen_content_mode                  = 2;
-    config_ptr->enable_hbd_mode_decision             = 1;
+    config_ptr->enable_hbd_mode_decision             = 1;//
```
Comment (xuguangxin) [inline Source/App/EncApp/EbAppConfig.c]: //

File: Source/Lib/Common/ASM_AVX2/convolve_avx2.c [hunk 2]
```diff
@@ -1635,6 +1635,115 @@ static INLINE __m256i calc_mask_avx2(const __m256i mask_base, const __m256i s0,
         _mm256_add_epi16(mask_base, _mm256_srli_epi16(diff, 4)));
     // clamp(diff, 0, 64) can be skiped for diff is always in the range ( 38, 54)
 }
+#if COMP_HBD
+
+void av1_build_compound_diffwtd_mask_highbd_avx2(
+    uint8_t *mask, DIFFWTD_MASK_TYPE mask_type, const uint8_t *src0,
+    int src0_stride, const uint8_t *src1, int src1_stride, int h, int w,
+    int bd) {
+  if (w < 16) {
```
Comment (xuguangxin) [inline Source/Lib/Common/ASM_AVX2/convolve_avx2.c]: According to https://github.com/OpenVisualCloud/SVT-AV1/blob/master/STYLE.md. we need to use 4 spaces indent.

File: Source/App/EncApp/EbAppConfig.c (hunk 1 above)
Comment (okhlif) [reply Source/App/EncApp/EbAppConfig.c]: Should be addressed Thank you

File: Source/Lib/Common/ASM_AVX2/convolve_avx2.c (hunk 2 above)
Comment (okhlif) [reply Source/Lib/Common/ASM_AVX2/convolve_avx2.c]: Should be addressed Thank you

**Comment by kylophone** [comment]
> hbd 0 (ii/Comp/ATB) ON vs OFF (reference)	58%
> hbd 1 (ii/Comp/ATB)  ON vs OFF	54%

I assume this means slower? 158% runtime?

**Comment by okhlif** [comment]
> > hbd 0 (ii/Comp/ATB) ON vs OFF (reference)	58%
> > hbd 1 (ii/Comp/ATB)  ON vs OFF	54%
> 
> I assume this means slower? 158% runtime?

Yes correct 58% slower

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 3]
```diff
@@ -36,6 +36,10 @@ extern "C" {
 
 #define IFS_8BIT_MD                  1
 
+#define COMP_HBD                     1
+#define INTERINTRA_HBD               1
+#define ATB_HBD                      1
+#define ATB_FIX                      1
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbDefinitions.h]: can you plz add description of each macro

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 4]
```diff
@@ -302,6 +317,25 @@ void inter_intra_search(
         list_idx1 = get_list_idx(rf[1]);
     assert(list_idx0 < MAX_NUM_OF_REF_PIC_LIST);
     assert(list_idx1 < MAX_NUM_OF_REF_PIC_LIST);
+//
+#if INTERINTRA_HBD
+
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: remove empty space

File: Source/Lib/Common/Codec/EbModeDecisionProcess.h [hunk 5]
```diff
@@ -309,7 +309,11 @@ extern "C" {
     EbBool                              md_staging_skip_rdoq;
 
 #if II_COMP_FLAG
+#if INTERINTRA_HBD
+    DECLARE_ALIGNED(16, uint8_t,      intrapred_buf[INTERINTRA_MODES][2 * 32 * 32]); //MAX block size for inter intra is 32x32
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecisionProcess.h]: can you plz explain "as comment"why *2?

**Comment by joelsole** [comment]
The performance is very nice, but a 54% encoding time increase is significant. It is worth digging deeper:
  - Could the complexity/performance trade-off for ii, comp, atb be obtained separately?
  - What is the corresponding complexity/performance trade-off for the 8-bit content in the 8-bit path?
  - Have you done any profiling to check where the added encoding time is coming from?
  - Could numbers for all resolutions be provided?

File: third_party/fastfeat/fast.c [hunk 6]
```diff
@@ -5,18 +5,18 @@
 
 xy* aom_fast9_detect_nonmax(const byte* im, int xsize, int ysize, int stride, int b, int* ret_num_corners)
 {
-	xy* corners;
-	int num_corners;
-	int* scores;
-	xy* nonmax;
+    xy* corners;
```
Comment (xuguangxin) [inline third_party/fastfeat/fast.c]: Sorry for missed this. 
This format change is not related to your main purpose.
